src/metabase/models/serialization/utils.clj

```diff
@@ -65,3 +68,152 @@
+
+    The default implementation assumes there will be many files:
+    - It calls (serdes-query this user-or-nil) to get the (lazy, reducible) set of entities to serialize.
+    - For each entity, it calls `(serdes-serialize-one e)` and expects a {file-path {...}} map.
+    - These maps are merged to form the returned map.
+
+    To serialize as a single file, override this to do something else. (Then `serdes-query` and `serdes-serialize-one`
+    are never called and can be stubs.)")
+
+  (serdes-query
+    [this user-or-nil]
+    "Performs the select query, possibly filtered, for all the entities of this type that should be serialized.
+    Returns the result of the db/select-reducible.
+    Defaults to a naive db/select-redcible for the entire model, ignoring the optional user argument.
+
+    Only called by `serdes-serialize-all`, either the default implementation or (optionally) by a custom one.
+    This exists to be an easy override point, when the default `serdes-serialize-all` is good but you need to filter
+    the returned set, eg. by dropping archived entities.")
+
+  (serdes-serialize-one
+    [this]
+    "Serializes a single entity into a YAML-friendly map, with its filename.
+    Returns a [file-name {entity map...}] pair.
+
+    Default implementation:
+    - Uses `\"$ENTITY_ID.yaml\"` or `\"$IDENTITY_HASH.yaml\"` as the filename.
+    - Returns the entity as a vanilla map with `:id` and any `:foo_at` fields removed.
+    - Adds the field `:type \"Collection\"` (etc.) to specify the model.
+
+    That suffices for a few simple entities, but most entities will need to override this.
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (serdes-upsert
+    [old-entity new-map]
+    "Given the original entity and the new deserialized map (not an IModel yet), upsert appdb to merge in the updated
+    entity.
+    Defaults to a naive update by primary key, using just the new map's values.")
+
+  (serdes-insert
+    [this new-map]
+    "Given the model (eg. Collection) and the new deserialized map (not an IModel), insert this new record into
+    the appdb.
+    Defaults to a straight db/insert! of this new map."))
+
+(defn- default-serialize-all [model user-or-nil]
+  (eduction (map serdes-serialize-one) (serdes-query model user-or-nil)))
+
+(defn- default-query [model _]
+  (db/select-reducible model))
+
+(defn serialize-one-plus [f]
+  (fn [entity]
+    [(format "%s%s%s.yaml" (name entity) File/separatorChar (or (:entity_id entity) (identity-hash entity)))
+     (-> (into {} entity)
+         (dissoc (models/primary-key entity))
+         (remove-timestamps)
+         (assoc :serdes-type (name entity))
+         f)]))
+
+(defn- default-upsert [old-entity new-map]
+  (db/update! old-entity (get old-entity (models/primary-key old-entity)) new-map))
+
+(defn- default-insert [model new-map]
+  (db/insert! model new-map))
```

Comment:
I wonder whether the toucans hook will affect these insert? Does these insert require all hooks to be `idempotent`?

For example in User/pre-insert hook, we hash the password. So does that mean if we deserialize it here, the inserted user will have password is the hash of a `hashed` password?

What could go wrong if we do `simple_insert`? which will by pass all hooks

Reply:
same qs for `db/update!`

Reply:
We almost certainly want `simple-insert`, actually. We don't want, for example, the import to generate `entity_id`s. I hadn't thought about that.

Is there a similar concern for an update?